File: backend/src/main/java/ma/investpro/repository/DepenseInvestissementRepository.java
package ma.investpro.repository;

import ma.investpro.entity.DepenseInvestissement;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.time.LocalDate;
import java.util.List;

@Repository
public interface DepenseInvestissementRepository extends JpaRepository<DepenseInvestissement, Long> {

    List<DepenseInvestissement> findByProjetId(Long projetId);

    List<DepenseInvestissement> findByFournisseurId(Long fournisseurId);

    List<DepenseInvestissement> findByConventionId(Long conventionId);

    List<DepenseInvestissement> findByAxeAnalytiqueId(Long axeAnalytiqueId);

    List<DepenseInvestissement> findByPaye(Boolean paye);

    List<DepenseInvestissement> findByDateFactureBetween(LocalDate dateDebut, LocalDate dateFin);

    @Query("SELECT d FROM DepenseInvestissement d WHERE " +
           "(:projetId IS NULL OR d.projet.id = :projetId) AND " +
           "(:fournisseurId IS NULL OR d.fournisseur.id = :fournisseurId) AND " +
           "(:conventionId IS NULL OR d.convention.id = :conventionId) AND " +
           "(:axeId IS NULL OR d.axeAnalytique.id = :axeId) AND " +
           "(:paye IS NULL OR d.paye = :paye) AND " +
           "(:dateDebut IS NULL OR d.dateFacture >= :dateDebut) AND " +
           "(:dateFin IS NULL OR d.dateFacture <= :dateFin)")
    List<DepenseInvestissement> findByFilters(
        @Param("projetId") Long projetId,
        @Param("fournisseurId") Long fournisseurId,
        @Param("conventionId") Long conventionId,
        @Param("axeId") Long axeId,
        @Param("paye") Boolean paye,
        @Param("dateDebut") LocalDate dateDebut,
        @Param("dateFin") LocalDate dateFin
    );

    @Query("SELECT SUM(d.montantTtc) FROM DepenseInvestissement d WHERE d.projet.id = :projetId")
    Double sumMontantTtcByProjet(@Param("projetId") Long projetId);

    @Query("SELECT SUM(d.montantTtc) FROM DepenseInvestissement d WHERE d.paye = true")
    Double sumMontantTtcPaye();

    @Query("SELECT SUM(d.montantTtc) FROM DepenseInvestissement d WHERE d.paye = false")
    Double sumMontantTtcNonPaye();
}

File: backend/src/main/kotlin/ma/investpro/repository/DepenseInvestissementRepository.kt
package ma.investpro.repository

import ma.investpro.entity.DepenseInvestissement
import ma.investpro.entity.StatutDepense
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.time.LocalDate

@Repository
interface DepenseInvestissementRepository : JpaRepository<DepenseInvestissement, Long> {

    /**
     * Recherche par numéro de facture
     */
    fun findByNumeroFacture(numeroFacture: String): DepenseInvestissement?

    /**
     * Recherche par fournisseur
     */
    fun findByFournisseurId(fournisseurId: Long): List<DepenseInvestissement>

    /**
     * Recherche par convention
     */
    fun findByConventionId(conventionId: Long): List<DepenseInvestissement>

    /**
     * Recherche par statut
     */
    fun findByStatut(statut: StatutDepense): List<DepenseInvestissement>

    /**
     * Recherche les dépenses payées
     */
    fun findByPayeTrue(): List<DepenseInvestissement>

    /**
     * Recherche les dépenses non payées
     */
    fun findByPayeFalse(): List<DepenseInvestissement>

    /**
     * Recherche par plage de dates
     */
    @Query("SELECT d FROM DepenseInvestissement d WHERE d.dateFacture BETWEEN :dateDebut AND :dateFin ORDER BY d.dateFacture DESC")
    fun findByDateFactureBetween(
        @Param("dateDebut") dateDebut: LocalDate,
        @Param("dateFin") dateFin: LocalDate
    ): List<DepenseInvestissement>

    /**
     * Recherche multi-critères
     */
    @Query("""
        SELECT d FROM DepenseInvestissement d
        WHERE (:fournisseurId IS NULL OR d.fournisseur.id = :fournisseurId)
        AND (:conventionId IS NULL OR d.convention.id = :conventionId)
        AND (:statut IS NULL OR d.statut = :statut)
        AND (:dateDebut IS NULL OR d.dateFacture >= :dateDebut)
        AND (:dateFin IS NULL OR d.dateFacture <= :dateFin)
        ORDER BY d.dateFacture DESC
    """)
    fun search(
        @Param("fournisseurId") fournisseurId: Long?,
        @Param("conventionId") conventionId: Long?,
        @Param("statut") statut: StatutDepense?,
        @Param("dateDebut") dateDebut: LocalDate?,
        @Param("dateFin") dateFin: LocalDate?
    ): List<DepenseInvestissement>

    /**
     * Total des dépenses par convention
     */
    @Query("""
        SELECT COALESCE(SUM(d.montantTtc), 0)
        FROM DepenseInvestissement d
        WHERE d.convention.id = :conventionId
    """)
    fun getTotalByConvention(@Param("conventionId") conventionId: Long): java.math.BigDecimal

    /**
     * Recherche étendue avec relations pré-chargées (évite N+1)
     */
    @Query("""
        SELECT d FROM DepenseInvestissement d
        LEFT JOIN FETCH d.fournisseur
        LEFT JOIN FETCH d.convention
        LEFT JOIN FETCH d.compteBancaire
        WHERE (:fournisseurId IS NULL OR d.fournisseur.id = :fournisseurId)
        AND (:conventionId IS NULL OR d.convention.id = :conventionId)
        AND (:statut IS NULL OR d.statut = :statut)
        AND (:paye IS NULL OR d.paye = :paye)
        AND (:dateDebut IS NULL OR d.dateFacture >= :dateDebut)
        AND (:dateFin IS NULL OR d.dateFacture <= :dateFin)
        ORDER BY d.dateFacture DESC
    """)
    fun searchFull(
        @Param("fournisseurId") fournisseurId: Long?,
        @Param("conventionId") conventionId: Long?,
        @Param("statut") statut: StatutDepense?,
        @Param("paye") paye: Boolean?,
        @Param("dateDebut") dateDebut: LocalDate?,
        @Param("dateFin") dateFin: LocalDate?
    ): List<DepenseInvestissement>

    /**
     * Toutes les dépenses avec relations pré-chargées (évite N+1)
     */
    @Query("""
        SELECT d FROM DepenseInvestissement d
        LEFT JOIN FETCH d.fournisseur
        LEFT JOIN FETCH d.convention
        LEFT JOIN FETCH d.compteBancaire
        ORDER BY d.dateFacture DESC
    """)
    fun findAllWithRelations(): List<DepenseInvestissement>

    /**
     * Comptage par statut payé
     */
    fun countByPaye(paye: Boolean): Long

    /**
     * Somme montant TTC par statut payé
     */
    @Query("SELECT COALESCE(SUM(d.montantTtc), 0) FROM DepenseInvestissement d WHERE d.paye = :paye")
    fun sumMontantTtcByPaye(@Param("paye") paye: Boolean): java.math.BigDecimal
}
